File: backend/src/models/Tarifa.ts
import { DataTypes, Model, Association } from 'sequelize';
import sequelize from '../config/database';
import { ITarifa } from '../types';
import Hotel from './Hotel';

class Tarifa extends Model<ITarifa> implements ITarifa {
  public id!: number;
  public hotel_id!: number;
  public data_checkin!: Date;
  public data_checkout!: Date;
  public preco!: number;
  public moeda?: string;
  public canal?: string;
  public tipo_quarto?: string;
  public readonly created_at!: Date;
  public readonly updated_at!: Date;

  // Associações
  public static associations: {
    hotel: Association<Tarifa, Hotel>;
  };
}

Tarifa.init(
  {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },
    hotel_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: Hotel,
        key: 'id'
      },
      onUpdate: 'CASCADE',
      onDelete: 'CASCADE'
    },
    data_checkin: {
      type: DataTypes.DATEONLY,
      allowNull: false,
      validate: {
        isDate: true,
        notEmpty: true
      }
    },
    data_checkout: {
      type: DataTypes.DATEONLY,
      allowNull: false,
      validate: {
        isDate: true,
        notEmpty: true,
        isAfterCheckin(value: string) {
          if (this.data_checkin && new Date(value) <= new Date(this.data_checkin)) {
            throw new Error('Data de checkout deve ser posterior à data de checkin');
          }
        }
      }
    },
    preco: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false,
      validate: {
        isDecimal: true,
        min: 0
      }
    },
    moeda: {
      type: DataTypes.STRING(3),
      allowNull: false,
      defaultValue: 'BRL',
      validate: {
        len: [3, 3],
        isUppercase: true
      }
    },
    canal: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    tipo_quarto: {
      type: DataTypes.STRING(255),
      allowNull: true
    }
  },
  {
    sequelize,
    modelName: 'Tarifa',
    tableName: 'tarifas',
    timestamps: true,
    underscored: true,
    indexes: [
      {
        fields: ['hotel_id']
      },
      {
        fields: ['data_checkin']
      },
      {
        fields: ['data_checkout']
      },
      {
        fields: ['preco']
      },
      {
        fields: ['canal']
      },
      {
        fields: ['hotel_id', 'data_checkin', 'data_checkout']
      }
    ]
  }
);

export default Tarifa;

File: backend/src/models/tarifa.py
from flask_sqlalchemy import SQLAlchemy
from datetime import datetime

db = SQLAlchemy()

class Tarifa(db.Model):
    __tablename__ = 'tarifas'
    
    id = db.Column(db.Integer, primary_key=True)
    hotel_id = db.Column(db.Integer, db.ForeignKey('hoteis.id'), nullable=False)  # NOVO
    data_checkin = db.Column(db.Date, nullable=False)
    data_checkout = db.Column(db.Date, nullable=False)
    preco = db.Column(db.Float, nullable=False)
    moeda = db.Column(db.String(3), default='BRL')
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    importacao_id = db.Column(db.Integer, db.ForeignKey('importacao_logs.id'), nullable=True)
    
    def to_dict(self):
        return {
            'id': self.id,
            'hotel_id': self.hotel_id,
            'data_checkin': self.data_checkin.isoformat() if self.data_checkin else None,
            'data_checkout': self.data_checkout.isoformat() if self.data_checkout else None,
            'preco': self.preco,
            'moeda': self.moeda,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'importacao_id': self.importacao_id
        }

class ImportacaoLog(db.Model):
    __tablename__ = 'importacao_logs'
    
    id = db.Column(db.Integer, primary_key=True)
    hotel_id = db.Column(db.Integer, db.ForeignKey('hoteis.id'), nullable=False)  # NOVO
    arquivo_nome = db.Column(db.String(255), nullable=False)
    total_registros = db.Column(db.Integer, default=0)
    registros_validos = db.Column(db.Integer, default=0)
    registros_erro = db.Column(db.Integer, default=0)
    status = db.Column(db.String(50), default='processando')
    detalhes_erro = db.Column(db.Text, nullable=True)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    
    # Relacionamento com tarifas
    tarifas = db.relationship('Tarifa', backref='importacao', lazy=True)
    
    def to_dict(self):
        return {
            'id': self.id,
            'hotel_id': self.hotel_id,
            'arquivo_nome': self.arquivo_nome,
            'total_registros': self.total_registros,
            'registros_validos': self.registros_validos,
            'registros_erro': self.registros_erro,
            'status': self.status,
            'detalhes_erro': self.detalhes_erro,
            'created_at': self.created_at.isoformat() if self.created_at else None
        }